sqlitecluster/SQLiteNode.cpp

```diff
@@ -74,6 +80,123 @@ SQLiteNode::~SQLiteNode() {
+                // loop over.
+                node._replicationCommands.wait(lock);
+                continue;
+            }
+        }
+
+        // We just got a new command.
+        auto& command = p.second;
+        if (SIEquals(command.methodLine, "BEGIN_TRANSACTION")) {
+            try {
+                // Convenience vars.
+                auto& peer = p.first;
+                auto& command = p.second;
+
+                // Wait for the DB to come up to date.
+                while (true) {
+                    unique_lock<recursive_mutex> lock(node._replicationCommands.getMutex());
+                    if (node._replicationThreadsShouldExit) {
+                        return;
+                    }
+                    if (command.calcU64("NewCount") == db.getCommitCount() + 1) {
+                        // We can unlock once we know our condition has passed, there's no race in case it changes after
+                        // we've checked it but before we wait again, as we wont wait. We do this before any DB operations
+                        // so that waiting on the DB can't block enqueueing new commands. But it's important that we do
+                        // hold this lock before checking the conditions that determine if we can proceed with our DB
+                        // operations.
+                        lock.unlock();
+
+                        // Importantly, we don't start our transaction until the previous transaction has fully completed.
+                        // In the next version, where we handle concurrent transactions, this will need to be able to start
+                        // regardless of whether the previous transaction has started.
+                        node.handleBeginTransaction(db, peer, command);
+                        break;
+                    } else {
+                        // Wait and then start from the beginning.
+                        node._replicationCommands.wait(lock);
+                    }
+                }
+
+                // Wait for a COMMIT or ROLLBACK.
+                while (true) {
+                    unique_lock<recursive_mutex> lock(node._replicationCommands.getMutex());
+                    if (node._replicationThreadsShouldExit) {
+                        db.rollback();
+                        return;
+                    }
+
+                    // Look up our hashes to see if we can COMMIT or ROLLBACK.
+                    bool commit = false;
+                    bool rollback = false;
+                    {
+                        lock_guard<mutex> hashlock(node._replicationHashMutex);
+                        commit = node._replicationHashesToCommit.count(command["NewHash"]);
+                        rollback = node._replicationHashesToRollback.count(command["NewHash"]);
+                    }
+
+                    // If we can't do either, keep waiting.
+                    if (!commit && !rollback) {
```

Comment:
If we get a commit and a rollback for the same transaction simultaneously, currently it'll get committed. I guess I don't care enough to handle this, it should never happen, it would be an error on LEADER's part to send these both.